bftclient/src/bft_client.cpp

```diff
@@ -156,42 +204,88 @@ Reply Client::send(const MatchConfig& match_config,
     if (auto reply = wait()) {
       expected_commit_time_ms_.add(
           std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start).count());
+      reply_certificates_.clear();
       return reply.value();
     }
     metrics_.retransmissions.Get().Inc();
   }
 
   expected_commit_time_ms_.add(request_config.timeout.count());
-  outstanding_request_ = std::nullopt;
+  reply_certificates_.clear();
   throw TimeoutException(request_config.sequence_number, request_config.correlation_id);
 }
 
+std::deque<Reply> Client::sendBatch(std::deque<WriteRequest>& write_requests, const std::string& cid) {
+  std::deque<Reply> replies;
+  std::chrono::milliseconds max_time_to_wait = 0s;
+  MatchConfig match_config = writeConfigToMatchConfig(write_requests.front().config);
+  auto batch_msg = initBatch(write_requests, cid, max_time_to_wait);
+  auto start = std::chrono::steady_clock::now();
+  auto end = start + max_time_to_wait;
+  while (std::chrono::steady_clock::now() < end && replies.size() != pending_requests_.size()) {
+    bft::client::Msg msg(batch_msg);  // create copy here due to the loop
+    if (primary_) {
+      communication_->send(primary_.value().val, std::move(msg));
+    } else {
+      std::set<bft::communication::NodeNum> dests;
+      for (const auto& d : match_config.quorum.destinations) {
+        dests.emplace(d.val);
+      }
+      communication_->send(dests, std::move(msg));
+    }
+
+    wait(replies);
+    metrics_.retransmissions.Get().Inc();
+  }
+  reply_certificates_.clear();
+  if (replies.size() == pending_requests_.size()) {
+    expected_commit_time_ms_.add(
+        std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start).count());
+    return replies;
+  }
+
+  expected_commit_time_ms_.add(max_time_to_wait.count());
+  throw BatchTimeoutException(cid);
+}
+
 std::optional<Reply> Client::wait() {
+  std::deque<Reply> replies;
+  wait(replies);
+  if (replies.empty()) {
+    static constexpr size_t CLEAR_MATCHER_REPLIES_THRESHOLD = 5;
+    // reply_certificates_ should hold just one request when using this wait method
```

Comment:
Irrelevant comment?

Reply:
Actually, this is relevant, this wait method should be called only to non-batched requests so the reply_certificates_ suppose to hold just one request.

Reply:
So, why do we have a 'for' loop a line below?

Reply:
Because I did it wrong, I'll fix it